playground/backend/cmd/server/controller_test.go

```diff
@@ -448,6 +453,58 @@ func TestPlaygroundController_GetRunError(t *testing.T) {
 	}
 }
 
+func TestPlaygroundController_Cancel(t *testing.T) {
```

Comment:
The code in processCode is becoming more complex (by necessity) - I think it would make sense to make sure we're not leaking goroutines. Do you think we could add checks for that?

I found this library that seems to check this properly: https://github.com/uber-go/goleak

(or it may not fit our use case - but worth considering)

Reply:
Yes, this is a good point. However, when I tried to use this library for our tests some tests were failed. It seems that during the processing of the tests, we are creating additional goroutines (for local storage and to send requests to the backend), so this library will find them as additional goroutines.

Reply:
Yes, this is a good point. Added using of this library for our controller's tests.